server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java

```diff
@@ -0,0 +1,308 @@
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal, Settings otherSettings) {
+        return Settings.builder()
+            .put(super.nodeSettings(nodeOrdinal, otherSettings))
+            // detect leader failover quickly
+            .put(LeaderChecker.LEADER_CHECK_RETRY_COUNT_SETTING.getKey(), 1)
+            .put(LeaderChecker.LEADER_CHECK_INTERVAL_SETTING.getKey(), "100ms")
+            .build();
+    }
+
+    public void testNodeJoinsCluster() {
+        internalCluster().startNodes(3);
+        String masterNodeName = internalCluster().getMasterName();
+        int numberOfNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName).state().getNodes().size();
+        int numberOfMasterNodesOriginallyInCluster = internalCluster().clusterService(masterNodeName)
+            .state()
+            .nodes()
+            .getMasterNodes()
+            .size();
+        List<String> namesOfDataNodesInOriginalCluster = getListOfDataNodeNamesFromCluster(masterNodeName);
+
+        // Attempt to add new node
+        String newNodeName = internalCluster().startDataOnlyNode();
+        ensureStableCluster(4);
+
+        // Assert the new data node was added
+        ClusterState state = internalCluster().clusterService(masterNodeName).state();
+        assertEquals(numberOfNodesOriginallyInCluster + 1, state.nodes().getSize());
+        assertEquals(namesOfDataNodesInOriginalCluster.size() + 1, state.nodes().getDataNodes().size());
+        assertEquals(numberOfMasterNodesOriginallyInCluster, state.nodes().getMasterNodes().size());
+
+        List<String> namesOfDataNodesInNewCluster = getListOfDataNodeNamesFromCluster(masterNodeName);
+        assertTrue(namesOfDataNodesInNewCluster.contains(newNodeName));
+        for (String nodeName : namesOfDataNodesInOriginalCluster) {
+            assertTrue(namesOfDataNodesInNewCluster.contains(nodeName));
+        }
+    }
+
+    @TestLogging(reason = "test includes assertions about logging", value = "org.elasticsearch.cluster.coordination.NodeJoinExecutor:INFO")
+    public void testNodeTriesToJoinClusterAndThenDifferentMasterIsElected() {
+        List<String> nodeNames = internalCluster().startNodes(3);
+        ensureStableCluster(3);
+        String originalMasterNodeName = internalCluster().getMasterName();
+        int numberOfNodesOriginallyInCluster = internalCluster().clusterService(originalMasterNodeName).state().getNodes().size();
+        int numberOfMasterNodesOriginallyInCluster = internalCluster().clusterService(originalMasterNodeName)
+            .state()
+            .nodes()
+            .getMasterNodes()
+            .size();
+        int numberOfDataNodesOriginallyInCluster = internalCluster().clusterService(originalMasterNodeName)
+            .state()
+            .nodes()
+            .getDataNodes()
+            .size();
+        // Determine upfront who we want the next master to be
+        final var newMasterNodeName = nodeNames.getFirst();
```

Comment:
This will sometimes re-elect the original master, causing the `assertNotEquals(originalMasterNodeName, internalCluster().getMasterName());` lower down to fail.

I'm not sure that's a bad thing - IMO it'd be simpler to have one test that covers both cases, but if you do want to keep them separate then you need to do something like this:

```suggestion
        final var newMasterNodeName = randomValueOtherThan(originalMasterNodeName, () -> randomFrom(nodeNames));
```